### PR TITLE
fix(Core/Spells): Always Select Correct Item When Updating Weapon Skill

### LINKED ISSUE #6158: Gaining Bow Skillups From Thrown Weapon Attacks When Swapping Mid-Attack
#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [x] enUS
- [ ] other (specify)

##### FACTION

- [x] Alliance
- [x] Horde

(Should happens to both factions)
##### CONTENT PHASE:

- [x] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

##### CURRENT BEHAVIOUR:

Original report: https://github.com/chromiecraft/chromiecraft/issues/739

Throwing a throwing knife and switching to a bow (before the knife hits) may cause to skill up in bows.

https://user-images.githubusercontent.com/22718174/120121414-49e2d600-c171-11eb-9329-fea3b0f30311.mp4

Demonstrating with a bow with no ammo equipped:

https://user-images.githubusercontent.com/81782124/120194832-46c51400-c25d-11eb-85c3-7f6096b2ac1b.mp4

##### EXPECTED BLIZZLIKE BEHAVIOUR:

Skill-up should be to throwing knives.

##### STEPS TO REPRODUCE THE PROBLEM:

1. Use the "Throw" spell (ID 2764).
2. Equip to a bow (after the knife has been thrown, before it has hit).
3. Hope for the skill-up.

##### EXTRA NOTES:

Might also happens with any other range weaponry.

Tester: couldn't make it happen with swapping bow -> thrown, but maybe I just wasn't fast enough.

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/3716ddf3e4e1ef138df334d01980fea019272001

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98815912-gaining-bow-skillups-from-thrown-weapon-attacks-when-swapping-mid-attack?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
Fixed #6158.

## Changes Proposed:
-  Always select item that has been used on spell cast. Important if skilling weapon up and swapping mid-attack weapons.

## Issues Addressed:

- Closes #6158
- Closes https://github.com/chromiecraft/chromiecraft/issues/739
- 
## Tests Performed:

- Tested ingame.

## How to Test the Changes:

1. Use the "Throw" spell (ID 2764).
2. Equip to a bow (after the knife has been thrown, before it has hit).
3. Hope for the skill-up.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by pklloveyou** [comment]
After consolidation #7102 
I'll test here again

**Review by Kitzunu — APPROVED** [review]

**Comment by pklloveyou** [comment]
Tested After consolidation #7102 Possible conflicts “SpellAuraEffects.cpp”

**Review by Helias — APPROVED** [review]

File: src/server/game/Entities/Player/PlayerUpdates.cpp [hunk 1]
```diff
@@ -951,6 +951,11 @@ void Player::UpdateWeaponSkill(Unit* victim, WeaponAttackType attType)
     uint32 weapon_skill_gain = sWorld->getIntConfig(CONFIG_SKILL_GAIN_WEAPON);
 
     Item* tmpitem = GetWeaponForAttack(attType, true);
+    if (item && item != tmpitem && !item->IsBroken())
```
Comment (pklloveyou) [inline src/server/game/Entities/Player/PlayerUpdates.cpp:954]: This caused the server to crash

Comment (pklloveyou) [reply src/server/game/Entities/Player/PlayerUpdates.cpp:954]: IsBroken

**Comment by pklloveyou** [comment]
I don't know how to reproduce the crash. Let others test it

**Comment by pklloveyou** [comment]
Wait for #7102 merging before testing

File: src/server/game/Entities/Player/PlayerUpdates.cpp (hunk 1 above)
Comment (UltraNix) [reply src/server/game/Entities/Player/PlayerUpdates.cpp:954]: No way it's crashing here.

**Comment by FrancescoBorzi** [comment]
what's the status of this?

**Comment by UltraNix** [comment]
@FrancescoBorzi To be merged. #7102 has nothing to do with it.

**Pull request merged**